test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/TestScenariosCrossProduct.java

```diff
@@ -0,0 +1,93 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package ir_framework.tests;
+
+import java.util.Set;
+
+import compiler.lib.ir_framework.*;
+import compiler.lib.ir_framework.shared.TestRunException;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @requires vm.debug == true & vm.compMode != "Xint" & vm.compiler2.enabled & vm.flagless
+ * @summary Test cross product scenarios with the framework.
+ * @library /test/lib /testlibrary_tests /
+ * @run driver ir_framework.tests.TestScenariosCrossProduct
+ */
+
+public class TestScenariosCrossProduct {
+    public static void main(String[] args) {
+        // Single set should test all flags in the set by themselves.
+        try {
+            TestFramework t1 = new TestFramework();
+            t1.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=51",
+                                               "-XX:TLABRefillWasteFraction=53",
+                                               "-XX:TLABRefillWasteFraction=64"));
+            t1.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            if (!e.getMessage().contains("The following scenarios have failed: #0, #1, #2")) {
+                throw e;
+            }
+        }
+
+        // The cross product of a set with one element and a set with three elements is three sets.
+        try {
+            TestFramework t2 = new TestFramework();
+            t2.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=53"),
+                                        Set.of("-XX:+UseNewCode", "-XX:+UseNewCode2", "-XX:+UseNewCode3"));
+            t2.start();
+            Asserts.fail("Should have thrown exception");
+        } catch (TestRunException e) {
+            if (!e.getMessage().contains("The following scenarios have failed: #0, #1, #2")) {
+                throw e;
+            }
+        }
+
+        // The cross product of two sets with two elements is four sets.
+        try {
+            TestFramework t3 = new TestFramework();
+            t3.addCrossProductScenarios(Set.of("-XX:TLABRefillWasteFraction=53", "-XX:TLABRefillWasteFraction=64"),
+                                        Set.of("-XX:+UseNewCode", "-XX:-UseNewCode"));
+            t3.start();
+            Asserts.fail("Should have thrown exception");
```

Comment:
Can I also do a Power-Set? We could do that with an empty string or null.
Can an entry also be multiple flags? Sometimes you need them in pairs.

Would be nice to have tests for that ;)